Fix Rainfall is handled differently between weather file values and Site:Precipitation water manager values #4153

yujiex opened the pull request:
- Fixes #4153
 
## PR Overview
The rain flag/indicator, `state.dataEnvrn->IsRain` indicates whether surfaces are wet. It determines
whether to use drybulb or wetbulb temperature, whether to reset exterior
convection coefficient, and affects subsequent heat balance calculations.

Previously the `IsRain` flag is only set by the weather file. The `Site:Precipitation` rain schedule has no effect on the rain flag. After the modification, the rain flag is set by `Site:Precipitation` if it is defined in the idf. Otherwise, it is set by epw rain fall amount. The idd field A6 of RunPeriod will be renamed to "Use Rain Indicators" in a future PR.
 - It also fixes a few issues in #9177  
   - Changed to use `TodayIsRain` instead of `IsRain` in lower bounding the precipitation on a rainy day labeled with the epw "Present Weather Codes".
   - Add the `state.dataWeatherManager->UseRainValues` check when lower bounding the precipitation on a rainy day.
   - Corrected the error of the `state.dataWeatherManager->IsRainThreshold` not dividing by the number of timesteps.

Currently, in this feature branch, the `IsRain` flag operates as is shown in the following pseudo-code, when state.dataWeatherManager->UseRainValues is true.

```
IsRain = TodayIsRain(timestep, hour), which gets value from TomorrowIsRain(timestep, hour)
if ((site:precipitation is defined) and (During Run Period)) {
  IsRain = (rainfall amount >= threshold)
}
```

`TomorrowIsRain` is determined by
- The first three digits of the PresWeathConds code ("Present Weather Codes" -- described in the
AuxiliaryPrograms document): if they are not 999, then the flag is set to true. This condition is usually not fulfilled. I sampled one weather file in each state. None of them met this condition.
- During design day simulation, this flag is controlled by the "rain indicator" in SizingPeriod:DesignDay

`rainfall amount` is computed from rainfall rate which gets value from
- `Site:Precipitation` if it is defined
- otherwise from liquid precipitation, which then gets data from the epw. However, if epw records 0 rain but the rain flag is true, then the liquid precipitation will be overwritten with 2mm per hour. This could happen when PreWeathConds codes are not 999 or during sizing period, when rain indicator is set to true but epw has rain fall = 0.

The following is the comparison of the "site rain status" between the current branch (fix) and the develop branch (develop) for July using 5ZoneWaterSystems.idf. We can see the current branch sets the status according to site precipitation depth, while the develop branch sets it according to liquid precipitation depth.
![rain_status_5ZoneWaterSystems](https://user-images.githubusercontent.com/13421161/155598040-ec8e100e-6369-45bc-a7a4-66991b24b822.png)

The following is the comparison of the "site rain status" between the current branch (fix) and the develop branch (develop) for July using EcoroofOrlando.idf
![rain_status_EcoroofOrlando](https://user-images.githubusercontent.com/13421161/155598155-4c25fc11-2cb5-4259-90f2-350e2fd36434.png)

## Regression diffs
As the CI only runs design-day simulations, some expected diffs are not showing up. Locally running regression tests forcing annual simulation leads to diffs in the following files. 

- EMSPlantLoopOverrideControl.idf
- 5ZoneWaterSystems.idf
- 5ZoneWaterSystems_NoSitePrec.idf
- EcoroofOrlando.idf
- EcoroofOrlando_NoSitePrec.idf

<img width="884" alt="image" src="https://user-images.githubusercontent.com/13421161/159378523-9a481f88-3874-4ccd-9f52-ab48f8fc20a7.png">

The diffs in 5ZoneWaterSystems.idf, EcoroofOrlando.idf, EMSPlantLoopOverrideControl.idf are due to the change in here:
```
            state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay);
            if (state.dataWaterData->RainFall.ModeID == DataWater::RainfallMode::RainSchedDesign && state.dataEnvrn->RunPeriodEnvironment) {
                // CurrentAmount unit: m
                state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount >= (state.dataWeatherManager->IsRainThreshold / 1000.0);
            }
```
The difference in the two "*_NoSitePrec" files are because of the change to the TodayIsRain flag.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
Initial commit

- Use RainFall.CurrentAmount to set isRain
- RunPeriod A6 to Use Rain Indicators as both sched and epw are involved
- Use TodayIsRain
- Add state.dataWeatherManager->UseRainValues check

yujiex commented:
Hi @mjwitte I created another PR for this rain flag fix.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2309,33 +2309,35 @@ namespace WeatherManager {
-            if ((state.dataEnvrn->LiquidPrecipitation == 0) && (state.dataEnvrn->IsRain)) {
-                state.dataEnvrn->LiquidPrecipitation = 1.5 / 1000.0;
+            if ((state.dataEnvrn->LiquidPrecipitation == 0) &&
+                // this is the Present Weather Codes field in epw. See column 1-3 in Table 2.16 Weather Codes Field Interpretation
```
It's confusing to have a comment in the middle of the if statement.

yujiex replied:
I just moved the comments above the if statement.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2309,33 +2309,35 @@ namespace WeatherManager {
+                // this is the Present Weather Codes field in epw. See column 1-3 in Table 2.16 Weather Codes Field Interpretation
+                (state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay)) &&
+                (state.dataWeatherManager->UseRainValues)) {
+                state.dataEnvrn->LiquidPrecipitation = 1.5 / 1000.0 / state.dataGlobal->TimeStepZone;
```
`NumOfTimeStepInHour` instead of `TimeStepZone` here. Also, the literal `1.5/1000` should be declared here as a `constexpr` variable, named something like `defaultLiquidPrecip`.

yujiex replied:
Right. I should have it divided by NumOfTimeStepInHours. I just added the defaultLiquidPrecip to hold the constant.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2309,33 +2309,35 @@ namespace WeatherManager {
-            }
-        }
 
         // if using epw precipitation, throw warning
```
Looking at this now, it would be clearer (and save a couple of if blocks) to move this (lines 2313 thru 2332) into `WaterManager::UpdatePrecipitation. Review that function for unnecessary if blocks as well.

yujiex replied:
I combined this with the if statement in UpdatePrecipitation

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2309,33 +2309,35 @@ namespace WeatherManager {
         state.dataEnvrn->TotalCloudCover = state.dataWeatherManager->TodayTotalSkyCover(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay);
         state.dataEnvrn->OpaqueCloudCover = state.dataWeatherManager->TodayOpaqueSkyCover(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay);
 
         if (state.dataWeatherManager->UseRainValues) {
-            state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(
-                state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay); //.or. LiquidPrecipitation >= .8d0)  ! > .8 mm
+            state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount > 0.0;
```
Is it ever possible for the weather file data to have a non-zero (small) value for LiquidPrecipitation but TodayIsRain is false? If so, this will change old behavior and flip the rain flag to true.

yujiex replied:
I'll double-check whether this could happen. The other behavior change would be if weather data has TodayIsRain = false but user-input site:precipitation defined non-zero rainfall value.

mjwitte replied:
>  The other behavior change would be if weather data has TodayIsRain = false but user-input site:precipitation defined non-zero rainfall value.

True, but that's the whole point of this PR in the first place. So that's a good change.

yujiex replied:
I just checked the TMY3 data in the EnergyPlus/weather folder. All of them have PresWeathCodes = 999999999. As a result, none of them can set TodayIsRain to true. I'll download some more weather data and see.

yujiex replied:
@mjwitte I also downloaded 36 more TMY3 whose states are not covered by the weather files in EnergyPlus folder. All of them also have PresWeathCodes = 999999999.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1442,7 +1442,7 @@ RunPeriod,
        \key No
        \default No
-  A6,  \field Use Weather File Rain Indicators
+  A6,  \field Use Rain Indicators
```
Given that I/O Freeze is almost frozen, the IDD change will likely have to wait until after v22.1 release. And it will need doc updates and a mention in the input rule file (because field name changes must be documented for epJSON).

yujiex replied:
I'll remove the I/O change for now.

mjwitte reviewed (commented):
This branch is behind develop.

yujiex committed:
Revert idd field renaming

yujiex committed:
Move monthly rain accumulator into UpdatePrecipitation

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex committed:
fix type

yujiex committed:
Unit test of rain flag set by site:precipitation

yujiex committed:
clang-format

yujiex commented:
The regression diffs here are because of the change here in the second line:         
```
         if (state.dataWeatherManager->UseRainValues) {
            state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount > 0.0;
        } else {
            state.dataEnvrn->IsRain = false;
        }
```
If changing the second line is reverted to before the change like this
`state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay);`
Then the regression diffs are gone.

yujiex commented:
Hi, @mjwitte would you mind taking another look at this branch and see if other changes are needed? I was wondering do you think the rain flag line here 
`state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount > 0.0;` 
should be changed to the following to also consider weather file PresWeathCodes as well?
`state.dataEnvrn->IsRain = (state.dataWaterData->RainFall.CurrentAmount > 0.0) || (state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay));`

mjwitte commented:
I think I see what's happening. CI runs design days only for all testfiles, except _5ZoneAirCooled_annual.idf  and _5ZoneAirCooled_LeapYear_annual.idf

For the diffs with EMSPlantLoopOverrideControl, EcoroofOrlando, and PythonPluginPlantLoopOverrideControl, my guess is that the Site:Precipitation is overriding the SizingPeriod:DesignDay Rain Indicator field with is set to No. So, the base runs have IsRain=false for the entire design day and these changes are flipping it to true for all or part of the day. Since SizingPeriod:DesignDay has a field for Rain, perhaps we should not reset it for design day simulations, which means it should use whatever is in `TodayIsRain`.

For the diffs with _5ZoneAirCooled_annual.idf  and _5ZoneAirCooled_LeapYear_annual.idf, I don't understand why that's changing - these don't have Site:Precipitation. I would add the site rain status as an output and compare develop vs this branch to see when it's changing and try to figure out why. Perhaps we're misunderstanding the incoming weather codes?

yujiex commented:
> _For the diffs with _5ZoneAirCooled_annual.idf and _5ZoneAirCooled_LeapYear_annual.idf, I don't understand why that's changing - these don't have Site:Precipitation. 

It's because it's using epw rain data and the `state.dataWaterData->RainFall.CurrentAmount > 0.0` triggered. Previously this rain flag is set solely by `state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay)`, and from all the weather files I sampled, this TodayIsRain is always false. After our change, the isRain flag is set by the `state.dataWaterData->RainFall.CurrentAmount > 0.0`

mjwitte commented:
So, you're saying with develop, Site Rain Status is always zero for 5ZoneAirCooled? Even though there are non-zero precipitation amounts?

yujiex commented:
> So, you're saying with develop, Site Rain Status is always zero for 5ZoneAirCooled? Even though there are non-zero precipitation amounts?

let me double-check

yujiex commented:
> state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay)

It's not all zero. I missed accounting for this part assigning to the TomorrowIsRain which later passes value to TodayIsRain
```
state.dataWeatherManager->TomorrowIsRain(ts, hour) = state.dataWeatherManager->TomorrowLiquidPrecip(ts, hour) >=
         (0.8 / double(state.dataGlobal->NumOfTimeStepInHour)); // Wthr%IsRain(Hour)
```

yujiex committed:
Change lowerbound of isRain to match TommorrowIsRain

yujiex commented:
Hi @mjwitte, I have inserted some comparison plots in the PR to explain the difference between the current fix and the develop branch.

mjwitte commented:
~@yujiex So maybe the rain flag is fine as-is (in develop)? I ran EcoRoofOrlando with and without Site:Precipitation with v22.1 IOFreeze('ish) which should be the same as develop. The total count of rain status flag (at timestep) is the same (oddly) but they are not at the same timesteps. So, apparently Site:Precipitation is impacting the rain flag. Where does that happen in the code?~

yujiex commented:
> rain status



> @yujiex So maybe the rain flag is fine as-is (in develop)? I ran EcoRoofOrlando with and without Site:Precipitation with v22.1 IOFreeze('ish) which should be the same as develop. The total count of rain status flag (at timestep) is the same (oddly) but they are not at the same timesteps. So, apparently Site:Precipitation is impacting the rain flag. Where does that happen in the code?

Hi @mjwitte Sorry for the late reply. I pulled develop branch today and tried adding an output variable to EcoroofOrlando.idf like this
`  Output:Variable,*,Site Rain Status,timestep;`
I compared the output "site rain status" with and without commenting out the site:precipitation object in the idf. It seems they are the same at timestep level as well 🤔.

mjwitte commented:
@yujiex Sorry, I must have messed up the comparison before. So the original premise here is correct, Rain Status is *not* affected by Site:Precipitation in develop.

The safe path forward is to fix that for weather period simulations with Site:Precipitation while maintaining no diffs for runs without Site:Precipitation. 

Since CI only runs design days except for  _5ZoneAirCooled_annual.idf and _5ZoneAirCooled_LeapYear_annual.idf (which don't have Site:Precipitation), we should expect no diffs from CI. We'll have to run locally to show diffs for 5ZoneWaterSystems and EcoRoofOrlando.

yujiex commented:
> Site:Precipitation while maintaining no diffs for runs without Site:Precipitation

Right, I'll double check in the local runs with annual simulation for the 5ZoneWaterSysterms as well.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex committed:
Rain flag follow dev branch with >=

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2334,8 +2330,7 @@ namespace WeatherManager {
         if (state.dataWeatherManager->UseRainValues) {
-            state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(
-                state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay); //.or. LiquidPrecipitation >= .8d0)  ! > .8 mm
+            state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount >= 0.8 / 1000.0 / double(state.dataGlobal->NumOfTimeStepInHour);
```
The last part of this is a fixed value once `NumOfTimeStepInHour` is known. There should be some place in WeatherManager with one-time inits, so this could be calculated once there and stored.

Also, based on the CI diffs, this is still overriding the SizingPeriod:DesignDay input for Rain. I guess I'm ok with that. Just be sure to explain that in the docs for Site:Precipitation.

yujiex replied:
Right right, the threshold for the rain flag should be stored.

I'll add documentation about the rain flag computed here overwriting the design day rain flag.

mjwitte replied:
Also, you may need to add some explanation about the default precip amount for the weather file and how all these parts interact.

yujiex replied:
I'll add some more explanation about the epw precipitation, user input site precipitation, default amount, and how they collectively affect the rain flag.

yujiex replied:
Hi @mjwitte, I felt maybe I should still keep the SizingPeriod:DesignDay rain indicator intact? As currently epw rainfall also controls the `IsRain` calculation (since we're using the `state.dataWaterData->RainFall.CurrentAmount`, which could come from both site:precipitation and epw), the inaccuracy in epw rain data might throw off the simulation? So I a check to make the overwritten of `IsRain` only happen during the actual simulation run periods.

yujiex commented:
Hi @mjwitte I ran some annual regression tests locally and saw some diffs in files like DynamicClothing.idf, where in the develop branch the following is not executed since the idf has Timestep = 1.
```
state.dataWeatherManager->TomorrowIsRain(ts, hour) = state.dataWeatherManager->TomorrowLiquidPrecip(ts, hour) >=
                                                                         (0.8 / double(state.dataGlobal->NumOfTimeStepInHour)); //Wthr%IsRain(Hour)

```
Do you happen to know where the `0.8 / double(state.dataGlobal->NumOfTimeStepInHour)` comes from and why there is a difference with timestep = 1? Thanks.

yujiex committed:
Lower bound following dev differ by timestep=1

yujiex committed:
Move monthly prec table col 1 calc back to weather manager

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex commented:
Hi @mjwitte I was going to move the calculation of `state.dataWaterData->RainFall.MonthlyTotalPrecInWeather` and `state.dataWaterData->RainFall.numRainyHoursInWeather` back to the weather manager, as they seem to produce the wrong aggregated monthly epw precipitation value. In the following comparison, the left is when they are in the weathermanager, and the right one is when they are in the watermanager.
<img width="1010" alt="image" src="https://user-images.githubusercontent.com/13421161/158562950-4a554c32-1d74-465e-8f81-ac24f9fc2ead.png">
Below is the precipitation from the USA_FL_Orlando.Intl.AP.722050_TMY3.stat. We can see the left column matches the .stat file and produces the correct aggregated epw precipitation.
<img width="1041" alt="image" src="https://user-images.githubusercontent.com/13421161/158563190-ab3754d8-6c35-4465-8b4b-fd984bc16ea9.png">

yujiex committed:
Add more conditions in isRain calculation

- When timestep > 1, use the IsRainThreshold
- Make user input and epw rain fall only impact IsRain during run period

yujiex committed:
Update unit test

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex committed:
Remove redundant thresholding

yujiex committed:
Update doc

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex commented:
Hi @mjwitte, the PR is ready for another look. Thanks :)

mjwitte commented:
@yujiex I've done some digging into the past to answer some of the questions here.
1. When https://github.com/NREL/EnergyPlus/issues/3573 was [addressed](https://github.com/NREL/EnergyPlusArchive/commit/15bb22b4d121c4a8c185243334b253bdbc8b6770) (back in 2012) the default precipitation amount when when the weather code says rain but the liquid precipitation is zero changed from 1.5mm to 2mm (per hour).
2. Then https://github.com/NREL/EnergyPlus/issues/4014 was [addressed](https://github.com/NREL/EnergyPlusArchive/commit/61fe67e2f42b5c21b0112ddbe38551748f60aca5) to fix some timestep problems.
3. So when no Site:Precipitation object is present (`state.dataWaterData->RainFall.ModeID == DataWater::RainfallMode::EPWPrecipitation`) everything is already in place to set rainfall to 2mm/hour is the epw rainflag is true (regardless of whether RunPeriod Use Weather File Rain Indicators is Yes or No).
4. So the new code added in #9177 that sets the rainfall to 1.5mm and throws a warning is not needed. It will never be reached.
5. This PR should reduce to:
    - [ ] Only alter the `state.dataWeatherManager->TodayIsRain` when Site:Precipitation is used.
    - [ ] Change docs to reflect 2mm/hr as the default precip when rain status is true and precip is zero (Auxiliary Program doc)
    - [ ] Change docs to show that rain status is set to true if precip is >0.8mm/hr   (Auxiliary Programs doc, and Site:Precipitation)

I will add some more comments in the code change page.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2317,10 +2318,13 @@ namespace WeatherManager {
+                state.dataEnvrn->LiquidPrecipitation = defaultLiquidPrecip / state.dataGlobal->NumOfTimeStepInHour;
                 ShowRecurringWarningErrorAtEnd(state,
                                                "Rain flag is on but precipitation depth in the weather file is missing or zero. Setting "
                                                "precipitation depth to 1.5 mm for this hour.",
```
All of this can be deleted, because the earlier weather processing has already set this to 2mm/hr [here](https://github.com/NREL/EnergyPlus/blob/f4bdf9af7765ddcefcb5da77d13acda9f6e3922d/src/EnergyPlus/WeatherManager.cc#L3157-L3160) and then adjusted for number of timesteps [here](https://github.com/NREL/EnergyPlus/blob/f4bdf9af7765ddcefcb5da77d13acda9f6e3922d/src/EnergyPlus/WeatherManager.cc#L3305-L3307).

yujiex replied:
Right, this will not be needed. I deleted it.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -3306,8 +3315,6 @@ namespace WeatherManager {
                         state.dataWeatherManager->LastHrLiquidPrecip * WtPrevHour + Wthr.LiquidPrecip(hour) * WtNow;
                     state.dataWeatherManager->TomorrowLiquidPrecip(ts, hour) /= double(state.dataGlobal->NumOfTimeStepInHour);
 
-                    state.dataWeatherManager->TomorrowIsRain(ts, hour) = state.dataWeatherManager->TomorrowLiquidPrecip(ts, hour) >=
-                                                                         (0.8 / double(state.dataGlobal->NumOfTimeStepInHour)); // Wthr%IsRain(Hour)
```
Keep this code as-is (do not delete).

yujiex replied:
Added it back

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2334,8 +2338,13 @@ namespace WeatherManager {
         state.dataEnvrn->OpaqueCloudCover = state.dataWeatherManager->TodayOpaqueSkyCover(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay);
 
         if (state.dataWeatherManager->UseRainValues) {
-            state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(
-                state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay); //.or. LiquidPrecipitation >= .8d0)  ! > .8 mm
```
Keep this code but add `if (state.dataWaterData->RainFall.ModeID == DataWater::RainfallMode::EPWPrecipitation)` and add a comment pointing to the code that sets `TomorrowIsRain` based on the 0.8mm/hr threshold.

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2334,8 +2338,13 @@ namespace WeatherManager {
         if (state.dataWeatherManager->UseRainValues) {
-            state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(
-                state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay); //.or. LiquidPrecipitation >= .8d0)  ! > .8 mm
+            // following the dev branch lower bound condition
+            if (state.dataWeatherManager->NumIntervalsPerHour == 1 && state.dataGlobal->NumOfTimeStepInHour > 1 &&
+                state.dataEnvrn->RunPeriodEnvironment) {
+                state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount >= state.dataWeatherManager->IsRainThreshold;
+            } else {
+                state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay);
+            }
```
This can reduce to:
```
elseif (state.dataWaterData->RainFall.ModeID == DataWater::RainfallMode::RainSchedDesign && state.dataEnvrn->RunPeriodEnvironment){
    state.dataEnvrn->IsRain = state.dataWaterData->RainFall.CurrentAmount >= state.dataWeatherManager->IsRainThreshold;
}

yujiex replied:
Changed the above two

mjwitte reviewed (commented):
@yujiex Some specific code comments.

mjwitte commented on doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:
```diff
@@ -1771,7 +1771,20 @@ \subsubsection{Inputs}\label{inputs-21-003}
+Precipitation includes both rain and the equivalent water content of snow.
+Precipitation is not yet described well enough in the many building weather data
+files. So this object can be used to provide the data using Schedule objects
+that define rates of precipitation in meters per hour. In 9.6 and later
```
In 22.1

yujiex replied:
Changed the version number

mjwitte commented on doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:
```diff
@@ -1771,7 +1771,20 @@ \subsubsection{Inputs}\label{inputs-21-003}
+Precipitation is not yet described well enough in the many building weather data
+files. So this object can be used to provide the data using Schedule objects
+that define rates of precipitation in meters per hour. In 9.6 and later
+versions, Site:Precipitation is also used in the determination of the
+rain indicator during the run period. As the rain fall rate is determined by
+Site:Precipitation when it is defined. Otherwise, it takes value from liquid
+precipitation from the epw. During the run period, when the rain rate is
+no less than 0.8 mm per hour, then the rain flag is overwritten to true.
+This overwritten occurs during the run period when there is only one weather
+file data periods and the number of timesteps in an hour is 1. During the sizing
+period, the rain indicator defined in the SizingPeriod:DesignDay will be used instead.
```
... versions, Site::Precipitation is also used to set the rain indicator. During a weather file simulation (RunPeriod), if Use Weather File Rain Indicators is Yes and the precipitation rate is no less than 0.8 mm per hour, then the rain flag is set to true. During a sizing simulation the rain indicator defined in SizingPeriod:DesignDay, SizingPeriod:WeatherFileDays, or SizingPeriod:WeatherFileConditionType will be used instead.

yujiex replied:
Changed the wording

mjwitte reviewed (commented):
Hmmm, not sure what will happen with SizingPeriod:WeatherFileConditionType and SizingPeriod:WeatherFileDays. Do you think users would want Site:Precipitation to set the rain flags for these? I guess we won't do that for now.

And some other doc edits.

yujiex committed:
Comments 03-21

- remove warning of 1.5mm threshold
- Change rain flag overwritten with CurrentAmount when site:prec exists
- put back 0.8mm thresholding of TomorrowIsRain

yujiex committed:
Merge remote-tracking branch 'origin/develop' into fixRainFlag

yujiex committed:
More doc fixes about rain indicator overwritten

yujiex committed:
Update unit test, change rain flag overwritten from ifelse to if

mjwitte commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -2317,27 +2318,19 @@ namespace WeatherManager {
         if (state.dataWeatherManager->UseRainValues) {
-            state.dataEnvrn->IsRain = state.dataWeatherManager->TodayIsRain(
-                state.dataGlobal->TimeStep, state.dataGlobal->HourOfDay); //.or. LiquidPrecipitation >= .8d0)  ! > .8 mm
-        } else {
-            state.dataEnvrn->IsRain = false;
```
@yujiex Looks like this else was deleted? Is this covered elsewhere?

yujiex replied:
In the header file, the IsRain is initialized to false. So I was thinking if I don't do anything, it should be false.

mjwitte replied:
But `UseRainValues` can change with each environment, so `IsRain` needs to be reset at least once per environment.

yujiex replied:
Right, I'll add it back.

yujiex committed:
Add back the else to explicitly set IsRain to false

mjwitte reviewed (approved):
Final answer! Assuming CI agrees (should be all green, no diffs in CI tests). 
Thanks @yujiex !

yujiex commented:
> Final answer! Assuming CI agrees (should be all green, no diffs in CI tests). Thanks @yujiex !

Thanks @mjwitte for all the comments and guidance :)
